pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -27,65 +27,133 @@ import (
 // Node is a set of constraints, compatible pods, and possible instance types that could fulfill these constraints. This
 // will be turned into one or more actual node instances within the cluster after bin packing.
 type Node struct {
-	Constraints         *v1alpha5.Constraints
+	v1alpha5.Constraints
 	InstanceTypeOptions []cloudprovider.InstanceType
 	Pods                []*v1.Pod
-	unschedulable       bool // if true, the node will accept no more pods
+
+	podResources    v1.ResourceList
+	daemonResources v1.ResourceList
+	daemons         []daemonInfo
+}
+
+type daemonInfo struct {
+	pod          *v1.Pod
+	requirements v1alpha5.Requirements
+	resources    v1.ResourceList
 }
 
-func NewNode(constraints *v1alpha5.Constraints, instanceTypeOptions []cloudprovider.InstanceType, pods ...*v1.Pod) *Node {
-	n := &Node{Constraints: constraints.DeepCopy()}
+func NewNode(constraints *v1alpha5.Constraints, daemons []*v1.Pod, instanceTypeOptions []cloudprovider.InstanceType, pods ...*v1.Pod) (*Node, error) {
+	n := &Node{
+		Constraints: *constraints.DeepCopy(),
+	}
+
+	for _, daemon := range daemons {
+		// skip any daemons that our provisioner configured taints would cause to not schedule
+		if err := n.Taints.Tolerates(daemon); err != nil {
+			continue
+		}
+		di := daemonInfo{
+			pod:          daemon,
+			requirements: v1alpha5.NewPodRequirements(daemon),
+			resources:    resources.RequestsForPods(daemon),
+		}
+		n.daemons = append(n.daemons, di)
+	}
+
```

Comment:
Logically, I'd expect something like
```
if err := n.Compatible(pod); err != nil {
  n.Add(pod)
}
```
Is there something I'm missing that prevents us from reusing the same mechanism? In my mind, the fact that you can pass pods into the constructor is just syntactic sugar, and the workflow should be identical if we decided to not provide that sugar.

Reply:
Didn't follow this, comment on wrong line?

Reply:
Or if in this case, n is a brand new node it won't work due to a peculiarity of how hostname topology spreads work.  We generate a node name requirement on the pod, it won't be compatible with the new node since it doesn't have a node name, but then when the requirements merge it's fine.  This will get taken care of in the next PR that handles affinity/anti-affinity correctly where we'll actually give the node the correct name.